include/swift/ABI/TaskStatus.h

```diff
@@ -154,12 +162,97 @@ class ChildTaskStatusRecord : public TaskStatusRecord {
   }
 };
 
+/// A status record which states that a task has a task group.
+///
+/// The child tasks are stored as an invasive single-linked list, starting
+/// from `FirstChild` and continuing through the `NextChild` pointers of all
+/// the linked children.
+///
+/// All children of the specific `Group` are stored "by" this record,
+/// so that they may be cancelled when this task becomes cancelled.
+///
+/// When the group exits, it may simply remove this single record from the task
+/// running it. As it has guaranteed that the tasks have already completed.
+///
+/// Group child tasks DO NOT have their own `ChildTaskStatusRecord` entries,
+/// and are only tracked by their respective `TaskGroupTaskStatusRecord`.
+class TaskGroupTaskStatusRecord : public TaskStatusRecord {
+  TaskGroup *Group;
+  AsyncTask *FirstChild;
+public:
+  TaskGroupTaskStatusRecord(TaskGroup *group)
+    : TaskStatusRecord(TaskStatusRecordKind::TaskGroup),
+      Group(group),
+      FirstChild(nullptr) {}
+
+  TaskGroupTaskStatusRecord(TaskGroup *group, AsyncTask *child)
+    : TaskStatusRecord(TaskStatusRecordKind::TaskGroup),
+      Group(group),
+      FirstChild(child) {}
+
+  TaskGroup* getGroup() const {
+    return Group;
+  }
+
+  /// Return the first child linked by this record.  This may be null;
+  /// if not, it (and all of its successors) are guaranteed to satisfy
+  /// `isChildTask()`.
+  AsyncTask *getFirstChild() const {
+    return FirstChild;
+  }
+
+  /// Attach the passed in `child` task to this group.
+  void attachChild(AsyncTask *child) {
+    assert(child->groupChildFragment());
+    assert(child->hasGroupChildFragment());
+    assert(child->groupChildFragment()->getGroup() == Group);
+
+    if (!FirstChild) {
+      // This is the first child we ever attach, so store it as FirstChild.
+      FirstChild = child;
+      return;
+    }
+
+    // We need to traverse the siblings to find the last one and add the child there.
```

Comment:
Is this ordering important?

Is there something which makes cancellation and this operation happen serially?

Reply:
Ah, you mean instead we just add at the head? Yes we could, the order does not matter.

> Is there something which makes cancellation and this operation happen serially?

Ouch oh no :-\ I missed that; It seems I would have to take the status lock for the duration of this function, right?